### ISSUE TITLE
Static deserialization issues

### ISSUE THREAD

**Issue by shawkins** [fabric8io/kubernetes-client#3972]
### Is your task related to a problem? Please describe

To eventually remove the static type registration from the KubernetesDeserializer we can't have static entry points that rely on this implicit class resolution.

### Describe the solution you'd like

Deprecate and provide alternatives to any static Serialization.unmarshall routines that don't supply the type.

**Comment by shawkins** [comment]
Specifically deprecate
```
UNMATCHED_FIELD_TYPE_MODULE
jsonMapper()
yamlMapper()
unmarshal(InputStream is)
unmarshal(InputStream is, Map<String, String> parameters)
unmarshal(InputStream is, ObjectMapper mapper)
unmarshal(InputStream is, ObjectMapper mapper, Map<String, String> parameters)
unmarshal(String str)
```

If we want the replacement to be available without completely building a client, then we'll need to add some method on the KubernetesClientBuilder.  Otherwise with the logic in #3966 it would be easier to obtain the deserializer from the client:

var serdes = client.serialization(); // type and method name tbd
serdes.unmarshal("something yaml");

Under the covers I'm assuming, but not knowledgeable enough with Jackson to known for sure, that we can associate an instance of the KubernetesDeserializer to use.

**Comment by shawkins** [comment]
At a minimum we would introduce here an api for accessing what are now static concerns of serialization:

```
public interface KubernetesSerialization {

  <T> T unmarshal(InputStream is);

  <T> T unmarshal(InputStream is, Map<String, String> parameters);

  <T> T unmarshal(String str);

  <T> T clone(T resource);

  void setLogUnmatchedFieldWarnings(boolean logWarnings);

  void setRestrictUnmatchedFieldToTemplates(boolean restrictToTemplates);

  default void registerCustomKind(String kind, Class<? extends KubernetesResource> clazz) {
    registerCustomKind(null, kind, clazz);
  }

  void registerCustomKind(String apiVersion, String kind, Class<? extends KubernetesResource> clazz);

  // optionally we could include pathutils methods

  String withoutRuntimeState(KubernetesResource object, Format format, boolean omitStatus);

  String jsonDiff(KubernetesResource current, KubernetesResource updated, boolean omitStatus);

}
```

Assuming that we want users to obtain this without creating a client, there would likely need to be another factory:  KubernetesSerialization.Factory or similar.  It would also be obtainable off of the client: client.getKubernetesSerialization()

On the public interface it doesn't seem advisable to directly expose the mapper instances.  Those would be available for internal use on the implementation class.

@manusa @rohanKanojia does this seem reasonable?

**Comment by shawkins** [comment]
After talking with @manusa the usage of registerCustomKind is more limited for the client than I thought.  We probably want to avoid using that internally or promoting that for usage by users.

This may require users to deal more with generic resources, which we'd like to offer a KubernetesResource.adapt(class) call.

This would also affect the Serialization.clone method.

**Comment by shawkins** [comment]
The next steps here is to remove all internal usage of KubernetesDeserializer.registerCustomKind - even if we have a non-static KubernetesDeserializer, that kind of side implicit side-effect is undesirable.   The expectation moving forward will be that the class is jandex discoverable.  In some situations if it is not, rather than seeing the typed result it would instead be a generickubenetesresource.  The places affected by this change will be: 
- direct untyped user calls to Serialization
- client.load - if the item is at a top level, the api server will be consulted to create the ResourceHandler.  If not found, then an exception will be thrown.
- transitive references to the type via KubernetesResource, or HasMetadata - under a template for example

All other kubernetesclient calls will function as expected because they are typed.

The only issue with KubernetesResource.adapt is the handling of mutability since the real type is not a view, but a converted instance.  It would be on the user to after making modifications, set the modified item back:

```
MyResource resource = something.getItem().adapt(MyResource.class) ;
// do something
something.setResource(resource);
```

**Comment by manusa** [comment]
I think there are still some tasks to be done in scope of this issue

**Comment by shawkins** [comment]
Marking as resolved by #4662 - we'll make further adjustments to the KubernetesSerialization from there.
